rllib/offline/offline_data.py

```diff
@@ -0,0 +1,152 @@
+import functools
+import logging
+import numpy as np
+from pathlib import Path
+import ray
+from typing import Any, Dict, List, Union
+
+from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
+from ray.rllib.core.columns import Columns
+from ray.rllib.env.single_agent_episode import SingleAgentEpisode
+from ray.rllib.utils.compression import unpack_if_needed
+from ray.rllib.utils.typing import EpisodeType
+
+logger = logging.getLogger(__name__)
+
+# TODO (simon): Implement schema mapping for users, i.e. user define
+# which row name to map to which default schema name below.
+SCHEMA = [
+    Columns.EPS_ID,
+    Columns.AGENT_ID,
+    Columns.MODULE_ID,
+    Columns.OBS,
+    Columns.ACTIONS,
+    Columns.REWARDS,
+    Columns.INFOS,
+    Columns.NEXT_OBS,
+    Columns.TERMINATEDS,
+    Columns.TRUNCATEDS,
+    Columns.T,
+    # TODO (simon): Add remove as soon as we are new stack only.
+    "agent_index",
+    "dones",
+    "unroll_id",
+]
+
+
+class OfflineData:
+    def __init__(self, config: Union[Dict[str, Any], AlgorithmConfig]):
+
+        self.config = config
+        self.is_multi_agent = (
+            config.is_multi_agent()
+            if isinstance(config, AlgorithmConfig)
+            else len(config.get("policies", {})) > 1
+        )
+        self.path = (
+            config.get("input_")
+            if isinstance(config.get("input_"), list)
+            else Path(config.get("input_"))
+        )
+        # Use `read_json` as default data read method.
+        self.data_read_method = config.get("data_read_method", "read_json")
+        self.compressed = config.get("compressed", False)
+        try:
+            self.data = getattr(ray.data, self.data_read_method)(self.path)
+            logger.info("Reading data from {}".format(self.path))
+            logger.info(self.data.schema())
+        except Exception as e:
+            logger.error(e)
+        self.batch_iterator = None
+
+    def sample(
+        self,
+        num_samples: int,
+        return_iterator: bool = False,
```

Comment:
Not sure, do I understand this correctly that we would expect, e.g.:

```
data = offline_data_group.sample(return_iterator=True)

# Then pass the iterator into LearnerGroup, which then handles the actual sharding?
# Or does OfflineDataGroup already shard (and return an iterator over the shards)?

learner_group.update(data=data)
```

??

Reply:
@sven1977 Yes, `OfflineData` does the sharding in such that it defines how many shards should be used (=`num_learners`). Then this number of `StreamingSplitIterator` objects are returned.

Reply:
Got it! Thanks for clarifying.